### ISSUE TITLE
Profile [profilename] does not exist. Using Cross account roles. credentials_source=EcsContainer

### ISSUE THREAD

**Issue by pawel-miszkurka-cko** [serverless/serverless#5474]
# This is a Bug Report

## Description
Hi, I am trying to use --aws-profile feature to deploy lambda functions using crossaccount role specified in aws/credentials file. 
* What went wrong?
sls deploy --aws-profile [profilename]

  Error --------------------------------------------------

  Profile [profilename] does not exist

* What did you expect should have happened?
Project should be deployed to corresponding account.
* What was the config you used?
My credentials file looks like this:
```
[playground]
role_arn=somecrossaccountrolearn
credential_source=EcsContainer

[dev]
role_arn=somecrossaccountrolearn
credential_source=EcsContainer

[test]
role_arn=somecrossaccountrolearn
credential_source=EcsContainer

[qa]
role_arn=somecrossaccountrolearn
credential_source=EcsContainer]

```
* What stacktrace or error message from your provider did you see?

```Serverless: Creating Stack...

  Error --------------------------------------------------

  Profile playground does not exist

     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.

  Stack Trace --------------------------------------------

Error: Profile [profilename] does not exist
    at Object.addProfileCredentials (/usr/lib/node_modules/serverless/lib/plugins/aws/provider/awsProvider.js:86:15)
    at AwsProvider.getCredentials (/usr/lib/node_modules/serverless/lib/plugins/aws/provider/awsProvider.js:310:12)
    at AwsProvider.request (/usr/lib/node_modules/serverless/lib/plugins/aws/provider/awsProvider.js:197:30)
    at createDeployment.then (/usr/lib/node_modules/serverless/lib/plugins/aws/deploy/lib/createStack.js:39:61)
From previous event:
    at AwsDeploy.create (/usr/lib/node_modules/serverless/lib/plugins/aws/deploy/lib/createStack.js:39:36)
From previous event:
    at AwsDeploy.BbPromise.bind.then.catch (/usr/lib/node_modules/serverless/lib/plugins/aws/deploy/lib/createStack.js:88:14)
From previous event:
    at AwsDeploy.createStack (/usr/lib/node_modules/serverless/lib/plugins/aws/deploy/lib/createStack.js:81:13)
From previous event:
    at Object.aws:deploy:deploy:createStack [as hook] (/usr/lib/node_modules/serverless/lib/plugins/aws/deploy/index.js:102:10)
    at BbPromise.reduce (/usr/lib/node_modules/serverless/lib/classes/PluginManager.js:390:55)
From previous event:
    at PluginManager.invoke (/usr/lib/node_modules/serverless/lib/classes/PluginManager.js:390:22)
    at PluginManager.spawn (/usr/lib/node_modules/serverless/lib/classes/PluginManager.js:408:17)
    at AwsDeploy.BbPromise.bind.then (/usr/lib/node_modules/serverless/lib/plugins/aws/deploy/index.js:95:48)
From previous event:
    at Object.deploy:deploy [as hook] (/usr/lib/node_modules/serverless/lib/plugins/aws/deploy/index.js:91:10)
    at BbPromise.reduce (/usr/lib/node_modules/serverless/lib/classes/PluginManager.js:390:55)
From previous event:
    at PluginManager.invoke (/usr/lib/node_modules/serverless/lib/classes/PluginManager.js:390:22)
    at PluginManager.run (/usr/lib/node_modules/serverless/lib/classes/PluginManager.js:421:17)
    at variables.populateService.then.then (/usr/lib/node_modules/serverless/lib/Serverless.js:157:33)
    at runCallback (timers.js:789:20)
    at tryOnImmediate (timers.js:751:5)
    at processImmediate [as _immediateCallback] (timers.js:722:5)
From previous event:
    at Serverless.run (/usr/lib/node_modules/serverless/lib/Serverless.js:144:8)
    at serverless.init.then (/usr/lib/node_modules/serverless/bin/serverless:43:50)
    at <anonymous>

  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com

  Your Environment Information -----------------------------
     OS:                     
     Node Version:           8.9.3
     Serverless Version:     
```

## Additional Data

* ***Serverless Framework Version you're using***: 1.32.0
* ***Operating System***: linux
* ***Stack Trace***:
* ***Provider Error messages***:

**Comment by pawel-miszkurka-cko** [comment]
> Does the `playground` profile exist in `~/.aws/credentials` or in `~/.aws/config`?

It exists in `~/.aws/credentials`.

**Comment by dschep** [comment]
k. thanks. Have you tried using `AWS_PROFILE=playground sls deploy`? (i didn't actually realize we had an --aws-profile option. I've always used the env var that `aws-sdk` recognizes)

**Comment by garethmcc** [comment]
I have used the `--aws-profile` extensively in the past but that is with credentials `aws_access_key` and `aws_secret_access_key` not `role_arn` and `credential_source`. Perhaps that option doesn't support usage outside of specifying the access key and secret access key?

**Comment by garethmcc** [comment]
And `--aws-profile` has existed for the last two years of me doing serverless development so its not all that new.

**Comment by pawel-miszkurka-cko** [comment]
Thanks, @garethmcc, I can also confirm that --aws-profile works well while using `aws_secret_key` and `aws_secret_access_key` however it does not help me to solve my problem.

**Comment by pawel-miszkurka-cko** [comment]
not a real solution but a workaround for now. In my case, I ended up running a bash script that fetches temp credentials and sets them as environment variables. 
Prerequisite for running the script:  **jq** installed
```
#!/bin/bash

CREDS=$(aws sts assume-role --role-arn $1 --role-session-name $2)
RETURN_VAL=$?

if [ "$RETURN_VAL" -eq "0" ]; then

   KEYID=`      echo $CREDS | jq -r '.Credentials.AccessKeyId'`
   SECRETKEY=`  echo $CREDS | jq -r '.Credentials.SecretAccessKey'`
   TOKEN=`      echo $CREDS | jq -r '.Credentials.SessionToken'`

   export AWS_ACCESS_KEY_ID=$KEYID
   export AWS_SECRET_ACCESS_KEY=$SECRETKEY
   export AWS_SESSION_TOKEN=$TOKEN
else
  echo "Failed to set credentials"
fi
```
As far as my research goes, serverless or aws js-sdk does not support my use case.

**Comment by ghinks** [comment]
I'm adding my 2 cents worth that when I use a simple export of the aws secret and key it does work but if I specify a profile  
```
provider:
  name: aws
  runtime: nodejs8.10
  profile: lamba
```

when these are specified in the ~/.aws/config and credentials the named profile is not found when using serverless.

**Comment by dschep** [comment]
@ghinks does it work if you set the aws-sdk's `AWS_SDK_LOAD_CONFIG` env var option?

**Comment by ghinks** [comment]
unfortunately not :(

**Comment by ghinks** [comment]
but many thanks for the tip :)

**Comment by ghinks** [comment]
I'm quite interested in trying to debug the issue. I'll look at the serverless code to see when/and/or what the order of precedence is as it happily picks up the exported key and secret but does not see the profile. 
Hints to where to look first accepted :)

**Comment by dschep** [comment]
Thanks for checking @ghinks, #5432 might give you some ideas of where to start looking. It's the PR that clued me into that env var seems related since it's loading a profile from config.

**Comment by brettswift** [comment]
Thanks for looking into this.  I'm trying to test a cross account lambda,  so if this solution works for `serverless invoke`  as well that would be awesome.

I have tested the `AWS_PROFILE` env var,  and it will override the `--aws-profile` option when using `serverless invoke`.   I'm not sure the `--aws-profile` is ever read with the invoke command actually.

If a branch becomes available I'll try to make time to test it.   Let me know :)

**Comment by bweigel** [comment]
I did some digging & debugging and the issue is due to the aws-js-sdk. At the moment the implementation for the shared ini file (i.e. `~/.aws/credentials`) only supports a `source_profile`, not a `credentials_source`.

See from [**shared_ini_file_credentials.js**](https://github.com/aws/aws-sdk-js/blob/3e10a5e45982f26b549699402859a99b2038b0cc/lib/credentials/shared_ini_file_credentials.js#L199) in function `loadRoleProfile`:

```js
  loadRoleProfile: function loadRoleProfile(creds, roleProfile, callback) {
...
    var self = this;
    var roleArn = roleProfile['role_arn'];
    var roleSessionName = roleProfile['role_session_name'];
    var externalId = roleProfile['external_id'];
    var mfaSerial = roleProfile['mfa_serial'];
    var sourceProfileName = roleProfile['source_profile'];

    if (!sourceProfileName) {
      throw AWS.util.error(
        new Error('source_profile is not set using profile ' + this.profile),
        { code: 'SharedIniFileCredentialsProviderFailure' }
      );
    }
...
```

**Comment by ghinks** [comment]
sincere thanks

**Comment by pmuens** [comment]
Closing since this is something which needs to be fixed in the AWS SDK codebase Serverless depends on...

**Comment by avegao** [comment]
Any update with this?

**Comment by vamcs** [comment]
Ok, I've found a solution to this at least on Windows.

If you're sure that you have the AWS folder in ``~/.aws`` make sure that your HOME and USERPROFILE environment variables point to the same location.

In our case, HOME was pointing to `C:\` and USERPROFILE to `C:\Users\<username>`.

After making them point to the same location and that location is where the `.aws` folder is, we could deploy using the `--aws-profile` flag correctly.

**Comment by AGnias47** [comment]
This was showing up for me because I was running ```serverless deploy``` within a script that I was running as ```sudo```. In that case, it couldn't find my profile because my account had the ```~/.aws/credentials``` profile, but the root user didn't. Not saying this shouldn't have been obvious, just wanted to post my feedback in case anyone else is banging their head on the same wall as me.

**Comment by dugarsumit** [comment]
While using ~/.aws/credentials file make sure `aws_secret_key` and `aws_secret_access_key` are written in lowercase.

**Comment by erdemsahinardentblue** [comment]
sudo or not sudo

I am getting this error

`rifaterdemsahin@Rifats-MacBook-Pro hello-world % sudo sls deploy --aws-profile erdem
Password:
 
  Error --------------------------------------------------
 
  Error: Profile erdem does not exist
      at Object.addProfileCredentials (/usr/local/lib/node_modules/serverless/lib/plugins/aws/provider/awsProvider.js:100:15)
      at AwsProvider.getCredentials (/usr/local/lib/node_modules/serverless/lib/plugins/aws/provider/awsProvider.js:379:12)
      at AwsCommon.validate (/usr/local/lib/node_modules/serverless/lib/plugins/aws/lib/validate.js:27:51)
      at AwsCommon.tryCatcher (/usr/local/lib/node_modules/serverless/node_modules/bluebird/js/release/util.js:16:23)
      at Promise._settlePromiseFromHandler (/usr/local/lib/node_modules/serverless/node_modules/bluebird/js/release/promise.js:547:31)
      at Promise._settlePromise (/usr/local/lib/node_modules/serverless/node_modules/bluebird/js/release/promise.js:604:18)
      at Promise._settlePromiseCtx (/usr/local/lib/node_modules/serverless/node_modules/bluebird/js/release/promise.js:641:10)
      at _drainQueueStep (/usr/local/lib/node_modules/serverless/node_modules/bluebird/js/release/async.js:97:12)
      at _drainQueue (/usr/local/lib/node_modules/serverless/node_modules/bluebird/js/release/async.js:86:9)
      at Async._drainQueues (/usr/local/lib/node_modules/serverless/node_modules/bluebird/js/release/async.js:102:5)
      at Immediate.Async.drainQueues [as _onImmediate] (/usr/local/lib/node_modules/serverless/node_modules/bluebird/js/release/async.js:15:14)
      at processImmediate (internal/timers.js:456:21)
      at process.topLevelDomainCallback (domain.js:137:15)`

**Comment by uqjhack1** [comment]
> I did some digging & debugging and the issue is due to the [AWS Javascript SDK](https://github.com/aws/aws-sdk-js), which serverless uses behind the scenes to get the credentials. At the moment the implementation for the shared ini file (i.e. `~/.aws/credentials`) only supports a `source_profile`, not a `credential_source`.
> 
> See from [**shared_ini_file_credentials.js**](https://github.com/aws/aws-sdk-js/blob/3e10a5e45982f26b549699402859a99b2038b0cc/lib/credentials/shared_ini_file_credentials.js#L199) in function `loadRoleProfile`:
> 
> ```js
>   loadRoleProfile: function loadRoleProfile(creds, roleProfile, callback) {
>     ...
>     var self = this;
>     var roleArn = roleProfile['role_arn'];
>     var roleSessionName = roleProfile['role_session_name'];
>     var externalId = roleProfile['external_id'];
>     var mfaSerial = roleProfile['mfa_serial'];
>     var sourceProfileName = roleProfile['source_profile'];
> 
>     if (!sourceProfileName) {
>       throw AWS.util.error(
>         new Error('source_profile is not set using profile ' + this.profile),
>         { code: 'SharedIniFileCredentialsProviderFailure' }
>       );
>     }
>    ...
> ```
> 
> Maybe one can try to do a workaround in the serverless code, but I think this is best addressed in the SDK itself. It is currently an open feature request there [aws/aws-sdk-js#1916](https://github.com/aws/aws-sdk-js/issues/1916).

After looking at this answer I just put this in the serverless.yml and worked as a charm.


provider:
  name: aws
  runtime: nodejs12.x
  source_profile: the-lambdas-profile
  ...

It looks like the serverless code is expecting a prop named "source_profile" and the doc says that you have to use the prop "profile"

Not sure if it is a Serverless concern to fix it in aws-cli, but It would be nice to fix the documentation so the doc and source code in serverless match.

check doc here
https://serverless.com/framework/docs/providers/aws/guide/credentials/

Thanks

**Comment by fabtheonly** [comment]
> I'm adding my 2 cents worth that when I use a simple export of the aws secret and key it does work but if I specify a profile
> 
> ```
> provider:
>   name: aws
>   runtime: nodejs8.10
>   profile: lamba
> ```
> 
> when these are specified in the ~/.aws/config and credentials the named profile is not found when using serverless.

thank you so much!
missing to disable this simple setting took me hours without any success. this finally solved the issue!

**Comment by MatteoGioioso** [comment]
Same error here, with the `aws-cli` works fine, however with `sls` it does not.

**Comment by micsafi** [comment]
Is this issue really solved? I am trying to accomplish the same cross-account scenario as the original author (@pawel-miszkurka-cko ), with using AWS profiles with only `role_arn` and `credential_source` configured. I am getting the same Profile 'xxx' does not exist error.

In my scenario I am running `sls` commands from Jenkins CI/CD pipeline which is hosted in a EC2 instance. Using access keys in such cross-account scenario would be really step backwards, since AWS provides the assume role functionality specifically for this kind of use cases.

**Comment by nocode99** [comment]
@micsafi I'm not sure what the previous issue one user was having but `AWS_SDK_LOAD_CONFIG=1` as an env var worked for me using a cross-account IAM role

**Comment by micsafi** [comment]
Yeah, after some additional research it seems that this is still a open upstream issue with the aws-sdk-js https://github.com/aws/aws-sdk-js/issues/1916. There is seems to a PR suggested, but it's not yet merged.

**Comment by BrutalSimplicity** [comment]
> @micsafi I'm not sure what the previous issue one user was having but `AWS_SDK_LOAD_CONFIG=1` as an env var worked for me using a cross-account IAM role

Yep, that also worked for me. It appears that when loading the credentials for the SDK it doesn't explicitly load the profiles from the config unless that is set. That behavior sounds unintended, but I'm not sure.

I believe the  culprit is here...

https://github.com/aws/aws-sdk-js/blob/c2ad3ef8fd486bb58edc225f74152dda51f9f2fe/lib/util.js#L959

That line only loads profiles from the config if the `AWS_SDK_LOAD_CONFIG` environment variable is set. It seems like odd behavior, but appears to be intentional based on this commit, https://github.com/aws/aws-sdk-js/commit/44b15949c58f5b969582fff97fe53c47c555dfa1.

/shrug

**Comment by Scout2012** [comment]
I had this same issue and was able to solve it without the need to set the `AWS_SDK_LOAD_CONFIG` environment variable. My issue was that my credentials file was setting the access key ID for aws in all capital letters, i.e. `AWS_ACCESS_KEY_ID=<access key>`, however my understanding is that the aws-sdk loads the keys from your credentials file as all lower case.
(reference: https://github.com/aws/aws-sdk-js/blob/c2ad3ef8fd486bb58edc225f74152dda51f9f2fe/lib/credentials/shared_ini_file_credentials.js#L138)

Before making all the variables lowercase, `this.accessKeyId` was evaluating as undefined, thus triggering the `Profile [profilename] does not exist` error inside of awsProvider.js. After the change, `this.accessKeyId` evaluated to the correct value and my profile was recognized with no further problems.

**TL;DR:** change your credential variables from all uppercase to all lowercase

This may or may not be the recommended fix, I personally don't see a problem with it. I just wanted to share what worked for me!

**Comment by jkodroff** [comment]
Since this still seems to be a problem affecting users, I would kindly ask that the maintainers re-open this issue, but maybe pin an explanation that it's an upstream issue and sls is waiting for AWS to publish a new version of the SDK in order for the issue to be properly resolved.

FWIW, `AWS_SDK_LOAD_CONFIG=1 sls deploy` did the job for me as a workaround.

**Comment by mmman64** [comment]
I'm also having issues with this. Although I can run `sls deploy` locally without issue, when running a build on AWS CodeBuild I run into "Error: Profile [profile_name] does not exist".

I've tried all the suggestions in this thread so far but no luck. I assumed the credentials for the aws profile I want to use would have been made available automatically but that doesn't seem to be the case. Am I missing something obvious here?

**Comment by mmman64** [comment]
Turns out this line in my build script was the culprit:

`'node_modules/.bin/sls' deploy -s $STAGE --aws-profile [my_profile]`

I removed `--aws-profile [my_profile]` and it finally worked!

**Comment by bcraft** [comment]
Currently having the same issue, I use a default profile with access key/secret key and then additional named profiles which only have a role arn and source (default).
Setting ` AWS_SDK_LOAD_CONFIG=1` improved things and the typescript compiler completed however the deployment fails with `"The security token included in the request is invalid"` which I cannot understand as I use this exact credential configuration for Terraform, CDK, boto3, aws-cli etc without any issues.

**Comment by ngduc** [comment]
For me, I have my "~/.aws/credentials" like this: (notice the "[default]" name)
```
[default]
aws_access_key_id = XXX
aws_secret_access_key = XXX
```
I ran the deploy command like this `serverless deploy --aws-profile default` ("default" is the profile name). Hope it helps!

**Comment by excenter** [comment]
Ran across a similar problem, I haven't implemented many of the fixes here, but we had success using the AWS CLI to STS assume role and immediately assume it in a temporary bash_env. (Headsup, this was designed for a CI/CD flow and wouldn't be a great fit for local development)
```
            aws sts assume-role \
            --role-arn << parameters.role-arn >> \
            --role-session-name << parameters.role-session-name >> \
            --query "Credentials.[AccessKeyId,SecretAccessKey,SessionToken]" \
            --output text | \
            awk '{print "export AWS_ACCESS_KEY_ID=\""$1"\"","\nexport AWS_SECRET_ACCESS_KEY=\""$2"\"","\nexport AWS_SESSION_TOKEN=\""$3"\""}' \
            >> $BASH_ENV && source $BASH_ENV

```

Having those 3 environment variables populated as the role you'd like to assume immediately before running performs subsequent actions as the assumed role.

**Comment by chrisgilbert** [comment]
I had the same problems as many others with this issue and burnt a lot of time on it.

Locally, I was able to get it to work OK, and couldn't figure out why it didn't work on jenkins build agents after a lot of trial and error.

The problem is that the aws-sdk v2 does not support using `credential_source=Ec2InstanceMetadata` along with `source_profile` and `role_arn`. There seems to have been support added to version 3, but this is not used by serverless yet, so that upgrade is not an option.

You can have a source profile which is referenced by a child profile, as long as the source profile sets aws credentials directly. You can't however assume a second role by using the IAM role attached to the EC2 instance or container.

You can also generate temporary credentials using the AWS CLI, and set the `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` directly in the environment. This solution would work so long as you are not using profile names directly in the serverless config.

For us, we are using `--aws-profile` and profile names in the serverless config, so the best solution I could find was to create a user that was able to assume the roles needed, and then set that up as the `source_profile`. This isn't exactly what I wanted (which was no shared credentials on build agents) but was a good enough interim step.

~/.aws/credentials
```
[assume-role-user]
aws_access_key_id = xxxxxx
aws_secret_access_key = xxxx
```

~/.aws/config
```
[profile serverless-deployments-dev]
role_arn = arn:aws:iam::ACCOUNT_ID:role/my-serverless-deployments-role
source_profile =  assume-role-user
role_session_name =  serverless-deploy
region = eu-west-2

...
```

To test it works in the CLI:
```
aws sts assume-role --profile serverless-deployments-dev
```

Then with serverless:
```
AWS_SDK_LOAD_CONFIG=1  sls deploy --aws-profile=serverless-deployments-dev
```

You can then have any number of profiles in the config file assume various roles in different accounts, but just one source profile and one shared key. The source profile will need something like this too, along with an appropriate policy for the role being assumed:

```
{
    "Version" : "2012-10-17",
    "Statement" : [
      {
        "Effect" : "Allow",
        "Action" : [
          "sts:AssumeRole",
          "sts:TagSession"
        ]
        "Resource" : [
          "arn:aws:iam::ACCOUNT_ID:role/my-serverless-deployments-role",
        ]
      }
    ]
  }
```
